crates/mockiato-codegen/src/generate/arguments.rs

```diff
@@ -1,61 +1,64 @@
-use super::constant::{arguments_ident, arguments_lifetime};
+use super::constant::{arguments_ident, arguments_lifetime, arguments_lifetime_as_generic_param};
+use super::generics::get_matching_generics_for_method_inputs;
 use super::lifetime_rewriter::{LifetimeRewriter, UniformLifetimeGenerator};
+use super::visibility::raise_visibility_by_one_level;
 use crate::parse::method_decl::MethodDecl;
 use crate::parse::method_inputs::MethodInputs;
+use crate::parse::trait_decl::TraitDecl;
 use proc_macro2::TokenStream;
 use quote::quote;
 use syn::visit_mut::visit_type_mut;
-use syn::Ident;
+use syn::{Generics, Ident};
 
 pub(crate) struct GeneratedArguments {
     pub(crate) output: TokenStream,
-    pub(crate) generics: TokenStream,
+    pub(crate) generics: Generics,
     pub(crate) ident: Ident,
 }
 
-pub(crate) fn generate_arguments(method_decl: &MethodDecl) -> GeneratedArguments {
+pub(crate) fn generate_arguments(
+    method_decl: &MethodDecl,
+    trait_decl: &TraitDecl,
+) -> GeneratedArguments {
     let arguments_ident = arguments_ident(&method_decl.ident);
 
     let mut lifetime_rewriter =
         LifetimeRewriter::new(UniformLifetimeGenerator::new(arguments_lifetime()));
     let arguments_fields = generate_arguments_fields(&mut lifetime_rewriter, &method_decl.inputs);
 
-    let generics = if lifetime_rewriter.generator.has_lifetimes() {
-        generics()
-    } else {
-        TokenStream::new()
-    };
+    let mut arguments_struct_generics =
+        get_matching_generics_for_method_inputs(&method_decl.inputs, &trait_decl.generics);
 
-    let debug_impl = generate_debug_impl(method_decl, &generics);
+    if lifetime_rewriter.generator.has_lifetimes() {
+        arguments_struct_generics
+            .params
+            .insert(0, arguments_lifetime_as_generic_param());
```

Comment:
If you are going to push thing regularly to the start of a `Vec`, consider using a `VecDeque` instead

Reply:
This is a struct from `syn` so I can't modify it.

Reply:
Actually, `syn` automatically puts lifetimes first when converting to a `TokenStream`:
https://docs.rs/syn/0.15.32/src/syn/generics.rs.html#862-897

I'm using `.push()` now: 59f61be